app/src/main/java/me/timpushkin/sgbmandroidapp/MainActivity.kt

```diff
@@ -0,0 +1,155 @@
+                                        openInputStream(right)?.use { rightStream ->
+                                            leftStream.readBytes() to rightStream.readBytes()
+                                        }
+                                    }
+                                }?.let {
+                                    mImages = it
+                                    imagesReady = true
+                                } ?: displayError("Cannot open images")
+                            },
+                            onRunClick = {
+                                mImages?.let { (left, right) ->
+                                    val (bitmap, time) = getDepthMap(left, right)
+                                    mDepthMap = bitmap
+                                    calculationTimeNanos = time
+                                } ?: throw IllegalStateException("Run clicked with no images")
+                            },
+                            onError = ::displayError
+                        )
+                    },
+                    backgroundColor = MaterialTheme.colors.background
+                ) { contentPadding ->
+                    Column(
+                        modifier = Modifier
+                            .fillMaxSize()
+                            .padding(contentPadding),
+                        verticalArrangement = Arrangement.SpaceBetween,
+                        horizontalAlignment = Alignment.CenterHorizontally
+                    ) {
+                        mDepthMap?.run {
+                            Image(
+                                bitmap = asImageBitmap(),
+                                contentDescription = "Depth map",
+                                modifier = Modifier.weight(1f)
+                            )
+                        }
+
+                        calculationTimeNanos?.let { time ->
+                            Text(text = "Calculation time: ${time * 1e-9} sec")
+                        }
+
+                        if (depthEstimatorReady && imagesReady) {
+                            ScaleFactorField { mDepthEstimator.setImageScaleFactor(it) }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private fun getDepthEstimator(uri: Uri): DepthEstimator? =
+        with(mStorageUtils) { uri.copyToCache(CACHED_PARAMS) }?.run { DepthEstimator(path) }
+
+    private fun getDepthMap(leftImage: ByteArray, rightImage: ByteArray): Pair<Bitmap, Long> {
+        val startTimeNanos = SystemClock.elapsedRealtimeNanos()
+        val depthArray = mDepthEstimator.estimateDepth(leftImage, rightImage)
+        val calculationTimeNanos = SystemClock.elapsedRealtimeNanos() - startTimeNanos
+        return depthArrayToBitmap(depthArray) to calculationTimeNanos
+    }
+
+    override fun onDestroy() {
+        super.onDestroy()
+
+        with(mStorageUtils) {
+            mImages?.let { (left, right) ->
```

Comment:
is onDestroy always called? is it great (enough time) to save cache here?

Reply:
Yep, I know it's not, but I initially thought only about persisting the images between events such as screen rotation, and here it is always called (or at least I think so...). Now it appeared to me that I can just use a `ViewModel` for this anyway, so no code in `onDestroy` anymore.